경북대 BE 김도현 - 1단계 카카오 로그인 구현

dozzzang opened the pull request:
안녕하세요 멘토님.
Step0에서 리뷰주신 부분 반영하여 Step1의 요구사항을 구현하였습니다.

## 구현사항
- 강의 자료를 기반으로 ``RestClient``를 활용해 인가 코드 → 액세스 토큰 → 사용자 정보 조회하는 3단계 OAuth 플로우를 구현했습니다.
- 카카오 클라이언트 키 등을 환경변수로 분리하여 Git에 노출되지 않도록 하였습니다.
- application.properties에서 ${KAKAO_CLIENT_ID} 형태로 최소한의 용도를 파악하기에 환경변수가 적절하다 판단하였습니다.

## 고민사항
- 기존에 구현한 User은 email로 회원가입/로그인을 수행하여 카카오 로그인을 통해 인증을 동기화하기 위해 ``accessToken``으로부터 email을 추출하고 싶었으나, 추가 기능을 신청해야하는 것을 깨닫고 계속 고민했으나 방법이 떠오르지 않아 우선은 "이메일 정보 없음"을 일관되게 반환하도록 설계하였습니다.

이 부분에 대해 멘토님의 조언을 구하고 싶습니다.
감사합니다!

dozzzang committed:
chore(step0) : README.MD 제목 수정

dozzzang committed:
chore(step0) : ErrorResponse final 필드로 수정

dozzzang committed:
refactor(step0) : option 최종 수량이 음수인 경우 구현

dozzzang committed:
chore(step0) : DELETE METHOD의 멱등성을 위해 예외 반환 삭제

dozzzang committed:
chore(step0) : 기존 카카오 상품명 메서드 삭제

dozzzang committed:
chore(step0) : transactional 어노테이션 추가

dozzzang committed:
docs : Step1 README.MD 작성

dozzzang committed:
feat : Step1 : 카카오 로그인 요구사항 구현

dozzzang committed:
test : KaKaoController 테스트코드 작성

jaenyeong commented on src/main/java/gift/auth/controller/KakaoController.java:
```diff
@@ -0,0 +1,70 @@
+          "error", "토큰 요청 실패",
+          "message", e.getMessage()
+      ));
+    }
```
🔴 
이렇게 처리하시면, 서버 에러가 발생해도 badRequest로 반환되지 않을까요..? ㅎㅎ 어떻게 생각하시나요?

dozzzang replied:
맞습니다. 4xx 에러 뿐만 아니라 5xx 에러 역시 발생할 수 있을 것 같습니다.
처음 답변주신대로 로그인을 1개의 API로 구성하여 예외 발생 시, login 페이지로 리다이렉트 될 수 있도록 수정하였습니다.

jaenyeong commented on src/main/java/gift/auth/controller/KakaoController.java:
```diff
@@ -0,0 +1,70 @@
+  }
+
+
+  @GetMapping("/kakao/login")
+  public ResponseEntity<Map<String, String>> getKakaoLoginUrl() {
+    String loginUrl = kakaoService.getKakaoLoginUrl();
+    return ResponseEntity.ok(Map.of("loginUrl", loginUrl));
+  }
+
+  @PostMapping("/kakao/token")
+  public ResponseEntity<Map<String, Object>> getToken
```
🟢 
가급적이면 Map이 아니라 ResponseDto를 구현하셔서 활용하시는 건 어떨까요? 😄

dozzzang replied:
현재 제가 구현한 dto는 ``KakaoTokenResponseDto``와  ``KakaoUserInfoDto``입니다.
``loginUrl``을 가져오는 메소드에서는 두 ``dto`` 모두 사용할 수 없으며 ``loginUrl``만을 필드로 가지는 ``dto``를 정의하기에는 불필요하다고 판단하였습니다.
1. ``DTO``는 복잡한 데이터 구조를 전송하거나 ``layer`` 간 데이터 교환을 위한 것이라고 알고 있습니다.
위 경우 ``loginUrl`` 하나 즉 단순한 문자열 하나이기 때문에 실용적이지 못하다고 생각했습니다.

2. 추후 카카오 API을 이용하는 코드가 추가된다 하더라도 ``loginUrl``만을 필드로 가지는 ``dto``는 재사용되지 않을 것이라 느꼈습니다!

이에 대해 멘토님의 조언을 구하고 싶습니다.

jaenyeong commented on src/main/java/gift/auth/service/KakaoService.java:
```diff
@@ -0,0 +1,78 @@
+
+  private final RestClient restClient;
+
+  public KakaoService() {
+    this.restClient = RestClient.create();
+  }
```
🔴 
현재 RestClient 인스턴스를 생성할 때 별도 설정을 안해주시는 것 같아요!
```java

static RestClient create() {
    return new DefaultRestClientBuilder().build();
}

@Override
public RestClient build() {
    ClientHttpRequestFactory requestFactory = initRequestFactory();
    UriBuilderFactory uriBuilderFactory = initUriBuilderFactory();

    HttpHeaders defaultHeaders = copyDefaultHeaders();
    MultiValueMap<String, String> defaultCookies = copyDefaultCookies();
    
    List<HttpMessageConverter<?>> converters =
		    (this.messageConverters != null ? this.messageConverters : initMessageConverters());
    
    return new DefaultRestClient(
		    requestFactory, this.interceptors, this.initializers,
		    uriBuilderFactory, defaultHeaders, defaultCookies,
		    this.defaultRequest,
		    this.statusHandlers,
		    converters,
		    this.observationRegistry, this.observationConvention,
		    new DefaultRestClientBuilder(this));
}


private ClientHttpRequestFactory initRequestFactory() {
    if (this.requestFactory != null) {
	    return this.requestFactory;
    }
    else if (httpComponentsClientPresent) {
	    return new HttpComponentsClientHttpRequestFactory();
    }
    else if (jettyClientPresent) {
	    return new JettyClientHttpRequestFactory();
    }
    else if (reactorNettyClientPresent) {
	    return new ReactorClientHttpRequestFactory();
    }
    else if (jdkClientPresent) {
	    // java.net.http module might not be loaded, so we can't default to the JDK HttpClient
	    return new JdkClientHttpRequestFactory();
    }
    else {
	    return new SimpleClientHttpRequestFactory();
    }
}

// 그리고 HttpComponentsClientHttpRequestFactory는 아래 필드가 있는데요
private long connectionRequestTimeout = -1L;
private long readTimeout = -1L;
```
위 값들을 그대로 사용한다면 어떤 일이 발생할 수 있을까요?

dozzzang replied:
멘토님께서 주신 코드를 인텔리제이에서 살펴봐도
```java
private long connectionRequestTimeout = -1L;
private long readTimeout = -1L;
```
에서의 -1L의 의도를 파악할 수 없었습니다.
Spring 공식 문서(https://docs.spring.io/spring-framework/reference/integration/rest-clients.html#rest-restclient-configuration) 에 따르면
For production use, you should configure timeouts and other connection parameters. 와 같은 문장이 있었고
구글링해보니,
현재 제 코드 방식인 ``RestClient.create()`` 대로면 카카오 API 서버에서 문제가 발생할 시 쓰레드가 무제한으로 대기하는 문제 즉, 타임아웃이 존재하지 않는 문제가 존재하였습니다. 이와 같은 상태는 좀비 상태로 메모리가 계속 누수될 수 있다고 생각합니다.

따라서 most comprehensive configuration options을 제공한다고 나와있는 ``HttpComponentsClientHttpRequest``을 사용하여
1. 카카오 서버에 연결하는 데 걸리는 타임아웃을 나타내는 ``setConnectTimeout``
2. 카카오 서버가 응답하는데 걸리는 타임아웃을 나타내는
``setReadTimeout``
3. 쓰레드 커넥션 풀에서 connection을 얻는데 걸리는 타임아웃을 나타내는 ``setConnectionRequestTimeout``
을 설정하는 방향으로 수정하였습니다.

각각의 값은 5초씩 두어 아무리 못해도 10초 내에는 응답(혹은 리다이렉트)처리 될 수 있도록 수정하였습니다. 타임아웃에 대한 시간 설정은 찾아보아도 어떻게 최적으로 설정할 수 있을지 모르겠어 임의로 판단하였습니다.

필요한 코드들을 알려주셔서 빠르게 찾아볼 수 있었습니다 감사합니다

jaenyeong reviewed (changes requested):
도현님 
몇가지 코멘트를 드렸는데 🔴 코멘트는 확인/수정 요청 드립니다 ㅎㅎ

그리고 지금 `ProductPageTest`, `WishPageTest`, `ProductIntegrationTest` 실패하는데, 이 부분도 확인 부탁드립니다!

그리고 `kakao/login` -> `kakao/token` -> `kakao/userinfo` 플로우는 잘 확인하였는데요!
가능하시다면 1개의 API로 카카오 로그인 후에 다시 되돌아 올 수 있는 형태로 플로우를 구성해 주시면 좋겠습니다 ㅎㅎ 그리고 기존에 로그인이 필요한 API를 그대로 호출할 때 문제 없도록 해당 액세스 토큰을 잘 심어 주시면 되겠습니다!

> 기존에 구현한 User은 email로 회원가입/로그인을 수행하여 카카오 로그인을 통해 인증을 동기화하기 위해 accessToken으로부터 email을 추출하고 싶었으나, 추가 기능을 신청해야하는 것을 깨닫고 계속 고민했으나 방법이 떠오르지 않아 우선은 "이메일 정보 없음"을 일관되게 반환하도록 설계하였습니다.

카카오 이메일을 잘 받아오실 수 있을 것 같은데 조금 더 확인해 보시고, 정 안된다면 테스트용 이메일을 하드코딩 하셔도 좋습니다, 해당 스텝에서 중요한건 Oauth 에 대한 이해와 외부 오픈 API를 사용할 때 주의점들을 잘 고민해서 보완하는 설계/구현을 하였는가이기 때문에요 ㅎㅎ

dozzzang committed:
test : kakao_client 누락 문제 해결

dozzzang committed:
feat : 카카오 OAuth 로그인 통합화
- 콜백을 이용하여 카카오 로그인 과정을 하나의 플로우로 구현하였습니다.
- 이메일을 앱의 필수 항목으로 설정하였습니다.
- 이로 인해, User DB와 연동하여 자동 로그인/회원가입(더미 비밀번호
  이용)을 구현하였습니다.

dozzzang committed:
refactor : 예외 반환 시 로그인 페이지로 리다이렉트

dozzzang committed:
refactor : restClient timeout 설정 추가

dozzzang committed:
test : API통합으로 인한 테스트코드 수정

jaenyeong commented on src/main/java/gift/auth/service/KakaoService.java:
```diff
@@ -0,0 +1,87 @@
+@Service
+public class KakaoService {
+
+  @Value("${kakao.client.id}")
+  private String clientId;
+
+  @Value("${kakao.redirect.url}")
+  private String redirectUrl;
```
🟢 
이것도 생성자 주입을 하신다면 final 선언할 수 있겠네요 ㅎㅎ

dozzzang replied:
말씀처럼 clientId와 redirectUrl은 변수명에서도 알 수 있듯 final로 불변을 보장해주는 것이 적절하다고 생각합니다.
현재는 MockitoBean 어노테이션을 통해 KakaoService를 사용하고 있어 문제 되지 않지만, 단위테스트가 필요한 경우에서 스프링에 의존하기 보다는 자바에 의존하게끔 생성자 주입 이용하는 것이 더 효율적일 수 있겠네요!
자꾸 final을 생각해보지 않게되네요..

jaenyeong commented on src/main/java/gift/auth/service/KakaoService.java:
```diff
@@ -0,0 +1,87 @@
+        .contentType(MediaType.APPLICATION_FORM_URLENCODED)
+        .body(body)
+        .retrieve()
+        .onStatus(HttpStatusCode::is4xxClientError, (request, responseEntity) -> {
+          throw new RuntimeException("토큰 요청 실패 - 4xx 오류. 카카오톡 메시지 전송 권한을 확인하세요.");
+        })
+        .onStatus(HttpStatusCode::is5xxServerError, (request, responseEntity) -> {
+          throw new RuntimeException("카카오 서버 오류 - 5xx 오류");
+        })
```
🟢 
여기도 그렇지만, 호출하는 컨트롤러에서도 예외를 잡아 따로 로그를 남기고 있진 않네요? ㅎㅎ(getUserInfo 도 그렇네요)
글로벌 익셉션 핸들러 쪽에서도 `RuntimeException` 또는 `Exception`을 잡진 않고 있어서 말씀 드려 봅니다~

jaenyeong commented on src/main/java/gift/auth/service/KakaoService.java:
```diff
@@ -0,0 +1,87 @@
+  private static final String KAKAO_AUTH_URL = "https://kauth.kakao.com";
+  private static final String KAKAO_TOKEN_URL = "https://kauth.kakao.com/oauth/token";
+
+  private final RestClient restClient;
+
+  public KakaoService() {
+    HttpComponentsClientHttpRequestFactory requestFactory = new HttpComponentsClientHttpRequestFactory();
+
+    requestFactory.setConnectTimeout(5000);
+    requestFactory.setReadTimeout(5000);
+    requestFactory.setConnectionRequestTimeout(5000);
+
+    this.restClient = RestClient.builder()
+        .requestFactory(requestFactory)
+        .build();
+  }
```
🟢 
현재 카카오 서비스에서 비즈니스로직 처리 외에 API를 호출하는 역할 정도로 쓰인 것 같아요 ㅎㅎ
그래서 서비스레이어보단 `KakaoOauthClient` 같은 네이밍으로 컴포넌트로 선언해서 사용되어도 괜찮을 것 같네요!

저라면 `KakaoService`에서 `MemberRepository`를 참조해 `KakaoOauthClient` 에서 받아온 값으로 고객 데이터를 조회 및 저장할 것 같은데 어떻게 생각하시나요? ㅎㅎ

dozzzang replied:
말씀하신대로  비즈니스 로직이 없고 단순 HTTP요청과 응답(카카오 API 호출)을 처리하는 로직을 ``KakaoOauthClient``을 ``@Component``를 이용하여 구현해 보았습니다.
``@Component``를 사용하는 이유는 ``KakaoService``에서 Client를 의존해야하기 때문에 스프링이 관리하도록 빈 설정을 해야하기 때문입니다.

이에 카카오 로그인의 비즈니스 로직 
즉, 카카오에서 이메일(사용자 정보)을 조회하고 
이를 통해 User DB에서 사용자를 조회 혹은 생성하고 
JWT 토큰까지 생성하는 로직은 카카오 로그인이라는 업무에 관한 도메인 비즈니스 로직이므로 서비스에 두었습니다. 

SRP에 더 적합한 구조인 것 같습니다!

jaenyeong commented on src/main/java/gift/auth/controller/KakaoController.java:
```diff
@@ -0,0 +1,68 @@
+package gift.auth.controller;
+
+
+import static gift.exception.ErrorCode.INTERNAL_SERVER_ERROR;
+
+import gift.auth.dto.KakaoTokenResponseDto;
+import gift.auth.dto.KakaoUserInfoDto;
+import gift.auth.service.KakaoService;
+import gift.exception.InternalServerException;
+import gift.user.JwtTokenProvider;
+import gift.user.entity.User;
+import gift.user.service.UserService;
+import jakarta.servlet.http.Cookie;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import java.util.Map;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.RestController;
```
🟢 
참조되지 않는 임포트문은 제거해주세요~

jaenyeong reviewed (changes requested):
도현님~
잘해 주셨는데, 현재 부팅 시 에러가 발생하는데요~
디스코드 디엠으로도 같은 말씀 남겨 놨습니다만

현재 실행 시 아래 에러로 인해 `KakaoService` 클래스의 빈이 생성되지 못하고 있는데요
`java.lang.NoClassDefFoundError: org/apache/hc/client5/http/classic/HttpClient`

확인 한 번 해주시고 말씀 부탁드리겠습니다~

dozzzang committed:
chore : RestClient 타임아웃 설정을 위한 httpclient5 의존성 추가

jaenyeong reviewed (approved):
도현님 카카오 SNS 로그인은 잘 구현해 주셨는데요!
`http://localhost:8080/kakao/login` 진입했을때 로그인 url을 반환해주기보다는 리다이렉션 처리 해주셔도 좋을 듯 하네요 ㅎㅎ

그리고 남겨드린 코멘트는 다음 스텝 진행 시 여유가 되신다면 참고 부탁드립니다 ㅎㅎ

> DTO는 복잡한 데이터 구조를 전송하거나 layer 간 데이터 교환을 위한 것이라고 알고 있습니다. 위 경우 loginUrl 하나 즉 단순한 문자열 하나이기 때문에 실용적이지 못하다고 생각했습니다. 추후 카카오 API을 이용하는 코드가 추가된다 하더라도 loginUrl만을 필드로 가지는 dto는 재사용되지 않을 것이라 느꼈습니다!

여기에 답변 드리면 단순히 재사용성만을 위해 DTO(또는 ResponseDto)를 선언하는 것이 아니다 입니다~
DTO는 상황에 따라 레이어마다 따로 두는 경우가 있고 레이어 간 이동 목적이 아닌 Controller에서 API 반환만을 위한 ResponseDto를 구현하기도 하는데 이 경우는 도현님 말씀처럼 재활용되지는 못하겠죠 ㅎㅎ DTO 등에서 Map 을 지양하는 이유는 여러가지가 있겠지만, 타입 안정성 측면과 그리고 검증/직렬화 측면에서 한계가 있고 가독성이 저하된다는 부분도 있습니다~ 그 외 문서화 등 기타 부분도 영향을 받기 때문에 가급적 맵을 사용하는 것은 지양하길 권해드리는 편입니다 ㅎㅎ 그래서 단순 재사용성만 보고 DTO 구현/사용 여부를 결정짓는 것이 아니다라고 말씀 드리고 싶네요~ 물론 너무 간단한 것조차 만드는 게 피로도가 있다라는 것엔 동의 합니다 ㅎㅎ
